packages/lesswrong/server/fmCrosspost/tokens.ts

```diff
@@ -0,0 +1,45 @@
+import { DatabaseServerSetting } from "../databaseSettings";
+import jwt, { VerifyErrors } from "jsonwebtoken";
+import { MissingSecretError } from "./errors";
+
+const crosspostSigningKeySetting = new DatabaseServerSetting<string|null>("fmCrosspostSigningKey", null);
+
+const getSecret = () => {
+  const secret = crosspostSigningKeySetting.get();
+  if (!secret) {
+    throw new MissingSecretError();
+  }
+  return secret;
+}
+const jwtSigningOptions = {
+  algorithm: "HS256",
+  expiresIn: "30m",
+} as const;
+
+export const signToken = <T extends {}>(payload: T): Promise<string> =>
+  new Promise((resolve, reject) => {
+    jwt.sign(payload, getSecret(), jwtSigningOptions, (err, token) => {
+      if (token) {
+        resolve(token);
+      } else {
+        reject(err);
+      }
+    });
+  });
+
+export const verifyToken = <T extends {}>(token: string, validator: (payload: T) => void): Promise<T> =>
+  new Promise((resolve, reject) => {
+    jwt.verify(token, getSecret(), (err: VerifyErrors | null, decoded?: T) => {
+      const payload = decoded as T;
+      if (payload) {
+        try {
+          validator(payload);
```

Comment:
I'd feel better if the validator were an actual type guard.  As an example:
```
export const validateConnectCrossposterPayload = (payload: unknown): payload is ConnectCrossposterPayload => {
  if (!('userId' in payload) || typeof payload.userId !== "string") {
    throw new MissingParametersError(["userId"], payload);
  }
  return true;
}
```

Then we can do:
```
export const verifyToken = <T extends {}>(token: string, validator: (payload: unknown) => payload is T): Promise<T> =>
  new Promise((resolve, reject) => {
    jwt.verify(token, getSecret(), (err: VerifyErrors | null, decoded?: unknown) => {
      if (decoded) {
        try {
          validator(decoded);
          resolve(decoded);
        } catch (e) {
          reject(e);
        }
      } else {
        reject(err);
      }
    });
  });
```

This lets the type guards themselves drive the inferred typing of the responses.  Of course, type guard functions aren't actually safer than raw type casts - they're still telling the type system that you've done your due diligence and are willing to pinky-swear to the validity of the data, even if you can't demonstrate it in a way the type system finds legible.  But they force co-location of the runtime type checking and the corresponding type assertion, which makes it much easier for us to notice mistakes, and remember to update both when updating either.

I realize this might look like a somewhat tendentious nitpick, but I'm thinking of this as a broader pattern across the codebase and want to push us in this general direction.

Reply:
Nice! Done!